src/Simulator/FluidSimulator.cpp

```diff
@@ -128,100 +128,58 @@ bool cFluidSimulator::IsHigherMeta(NIBBLETYPE a_Meta1, NIBBLETYPE a_Meta2)
 
 
 
-// TODO Not working very well yet :s
-Direction cFluidSimulator::GetFlowingDirection(int a_X, int a_Y, int a_Z, bool a_Over)
+Vector3f cFluidSimulator::GetFlowingDirection(int a_X, int a_Y, int a_Z)
 {
 	if (!cChunkDef::IsValidHeight(a_Y))
 	{
-		return NONE;
+		return {};
 	}
+
 	BLOCKTYPE BlockID = m_World.GetBlock(a_X, a_Y, a_Z);
-	if (!IsAllowedBlock(BlockID))  // No Fluid -> No Flowing direction :D
-	{
-		return NONE;
-	}
 
-	/*
-	Disabled because of causing problems and being useless atm
-	char BlockBelow = m_World.GetBlock(a_X, a_Y - 1, a_Z);  // If there is nothing or fluid below it -> dominating flow is down :D
-	if ((BlockBelow == E_BLOCK_AIR) || IsAllowedBlock(BlockBelow))
+	if (!IsAllowedBlock(BlockID))  // No Fluid -> No Flowing direction :D
 	{
-		return Y_MINUS;
+		return {};
 	}
-	*/
 
-	NIBBLETYPE LowestPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z);  // Current Block Meta so only lower points will be counted
-	int X = 0, Z = 0;  // Lowest Pos will be stored here
+	NIBBLETYPE CentralPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z) & 0x07;
+	NIBBLETYPE LevelPoint[4];
 
-	if (IsAllowedBlock(m_World.GetBlock(a_X, a_Y + 1, a_Z)) && a_Over)  // check for upper block to flow because this also affects the flowing direction
+	// blocks around the checking pos
+	Vector3i Points[]
 	{
-		return GetFlowingDirection(a_X, a_Y + 1, a_Z, false);
-	}
+		{ a_X + 1, a_Y, a_Z },
+		{ a_X, a_Y, a_Z + 1 },
+		{ a_X - 1, a_Y, a_Z },
+		{ a_X, a_Y, a_Z - 1 }
+	};
 
-	std::vector< Vector3i * > Points;
-
-	Points.reserve(4);  // Already allocate 4 places :D
-
-	// add blocks around the checking pos
-	Points.push_back(new Vector3i(a_X - 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X + 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z + 1));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z - 1));
-
-	for (auto itr = Points.cbegin(), end = Points.cend(); itr != end; ++itr)
+	for (unsigned long i = 0; i < ARRAYCOUNT(LevelPoint); i++)
 	{
-		Vector3i * Pos = (*itr);
-		auto PosBlockID = m_World.GetBlock(Pos->x, Pos->y, Pos->z);
-		if (IsAllowedBlock(PosBlockID))
+		BlockID = m_World.GetBlock(Points[i].x, Points[i].y, Points[i].z);
+
+		if (IsAllowedBlock(BlockID))
 		{
-			NIBBLETYPE Meta = m_World.GetBlockMeta(Pos->x, Pos->y, Pos->z);
-
-			if (Meta > LowestPoint)
-			{
-				LowestPoint = Meta;
-				X = Pos->x;
-				Z = Pos->z;
-			}
+			LevelPoint[i] = 0x07 & m_World.GetBlockMeta(Points[i].x, Points[i].y, Points[i].z);
```

Comment:
This calculation doesn't work properly with falling water (bit 0x08 set). That needs a special case.